icechunk/src/store.rs

```diff
@@ -337,6 +339,57 @@ impl Store {
         }
     }
 
+    pub async fn delete_dir(&self, prefix: &str) -> StoreResult<()> {
+        if self.read_only().await {
+            return Err(StoreError::ReadOnly);
+        }
+
+        // TODO: Handling preceding "/" is ugly!
+        let path = format!("/{}", prefix.trim_start_matches("/"))
+            .try_into()
+            .map_err(|_| StoreError::BadKeyPrefix(prefix.to_owned()))?;
+        let res = {
+            let guard = self.session.read().await;
+            let node = guard.get_node(&path).await;
+            match node {
+                Ok(NodeSnapshot {
+                    node_data: NodeData::Group, path: node_path, ..
+                }) => Ok((NodeType::Group, node_path)),
+                Ok(NodeSnapshot {
+                    node_data: NodeData::Array(..),
+                    path: node_path,
+                    ..
+                }) => Ok((NodeType::Array, node_path)),
+                Err(err) => Err(err),
+            }
+        };
+
+        match res {
+            Ok((NodeType::Group, node_path)) => {
+                let mut guard = self.session.write().await;
+                Ok(guard.deref_mut().delete_group(node_path).await?)
+            }
+            Ok((NodeType::Array, node_path)) => {
+                let mut guard = self.session.write().await;
+                Ok(guard.deref_mut().delete_array(node_path).await?)
+            }
+            _ => {
```

Comment:
I think we need to match only for the `Err(NotFound)` (pseudocode)? Other types of errors (imagine missing credentials, etc) need to be bubbled up immediately.